### PR TITLE
migrations: Schematics for `TransferState`, `StateKey` and `makeStateKey` migration

### PR DESCRIPTION
**Pull request by JeanMeche**
These 3 classes have been moved from platform-browser to core by #49563

Should this PR also provide a tslint rule for g3  ?

### CODE REVIEW

**Comment by alan-agius4** [comment]
> Should this PR also provide a tslint rule for g3 ?

No, a migration is not necessary for G3 as these APIs are not used internally.

**Comment by alan-agius4** [comment]
@JeanMeche is this ready for review?

**Comment by JeanMeche** [comment]
@alan-agius4  Yes, I fixed the conflicts.

File: packages/core/schematics/migrations.json [hunk 1]
```diff
@@ -9,6 +9,11 @@
       "version": "16.0.0",
       "description": "In Angular version 15.2, the guard and resolver interfaces (CanActivate, Resolve, etc) were deprecated. This migration removes imports and 'implements' clauses that contain them.",
       "factory": "./migrations/guard-and-resolve-interfaces/bundle"
+    },
+    "migration-v16-transfer-state": {
+      "version": "16.0.0",
+      "description": "Since Angular v16, `TransferState`, `makeStateKey`, `StateKey` has moved to the core package.",
```
Comment (alan-agius4) [inline packages/core/schematics/migrations.json:15]: NIT, the description should be what the migration does.

```suggestion
      "description": "Replace `TransferState`, `makeStateKey`, `StateKey` imports from `@angular/platform-browser` to `@angular/core`.",
```

File: packages/core/schematics/migrations/transfer-state/README.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
+## DefaultIterableDiffer migration
```
Comment (alan-agius4) [inline packages/core/schematics/migrations/transfer-state/README.md:1]: Read me seems to be copy and pasted

File: packages/core/schematics/migrations/transfer-state/index.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+import {DirEntry, Rule, UpdateRecorder} from '@angular-devkit/schematics';
+import ts from 'typescript';
+
+import {findImportSpecifier} from '../../utils/typescript/imports';
+
+const importsToReplace = ['makeStateKey', 'StateKey', 'TransferState'];
+
+function* visit(directory: DirEntry): IterableIterator<ts.SourceFile> {
+  for (const path of directory.subfiles) {
+    if (path.endsWith('.ts') && !path.endsWith('.d.ts')) {
```
Comment (crisbeto) [inline packages/core/schematics/migrations/transfer-state/index.ts:17]: Should use the existing `canMigrateFile` helper.

File: packages/core/schematics/migrations/transfer-state/index.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+          recorder.remove(index, length).insertLeft(index, newImports);
+        } else {
+          // Remove '@angular/platform-browser' import
+          const index = -platformBrowserImport.getFullStart();
+          const length = platformBrowserImport.getFullWidth();
+          recorder.remove(index, length);
+        }
+
+        // Import XXX from @angular/core
+        const coreImport = findImportDeclaration('@angular/core', allImportDeclarations);
```
Comment (crisbeto) [inline packages/core/schematics/migrations/transfer-state/index.ts:95]: There's an `ImportManager` utility that can be used to add new imports.

Comment (JeanMeche) [reply packages/core/schematics/migrations/transfer-state/index.ts]: Hi Kris, thank you for you feedback, I've migrated to using the change tracker (and ImportManager) but I'm not quite satisfied with the result. 

* `ImportManager` was creating unique idenfiers, I had to add an option to force the name. 
* I had to reset the changeTracker or else the imports were being added. 

Let me know what you think of my changes and thank you for your time.

File: packages/core/schematics/migrations.json [hunk 5]
```diff
@@ -9,6 +9,11 @@
       "version": "16.0.0",
       "description": "In Angular version 15.2, the guard and resolver interfaces (CanActivate, Resolve, etc) were deprecated. This migration removes imports and 'implements' clauses that contain them.",
       "factory": "./migrations/guard-and-resolve-interfaces/bundle"
+    },
+    "migration-v16-transfer-state": {
+      "version": "16.0.0",
+      "description": "Replace `TransferState`, `makeStateKey`, `StateKey` imports from `@angular/platform-browser` to `@angular/core`.",
```
Comment (AndrewKushnir) [inline packages/core/schematics/migrations.json:15]:
```suggestion
    "migration-v17-transfer-state": {
      "version": "17.0.0",
      "description": "Updates `TransferState`, `makeStateKey`, `StateKey` imports from `@angular/platform-browser` to `@angular/core`.",
```

File: packages/core/schematics/migrations/transfer-state/README.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
+## TransferState, makeStateKey and StateKey migration
+
+`TransferState`, `makeStateKey`, `StateKey` has been moved from `@angular/platform-browser` to `@angular/core`
+
+This migration will replace `TransferState`, `makeStateKey`, `StateKey` imports from `@angular/platform-browser` to `@angular/core`.
```
Comment (AndrewKushnir) [inline packages/core/schematics/migrations/transfer-state/README.md:5]:
```suggestion
## Import TransferState-related symbols from `@angular/core`

The following symbols were moved from `@angular/platform-browser` to `@angular/core`:

* `TransferState`
* `makeStateKey`
* `StateKey`

This migration updates symbol imports to use `@angular/core`.
```

File: packages/core/schematics/migrations/transfer-state/utils.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import ts from 'typescript';
+
+import {ChangeTracker} from '../../utils/change_tracker';
+import {getImportSpecifiers, removeSymbolFromNamedImports} from '../../utils/typescript/imports';
+import {closestNode} from '../../utils/typescript/nodes';
+
+export const exposedInterfaces = new Set(['makeStateKey', 'StateKey', 'TransferState']);
```
Comment (AndrewKushnir) [inline packages/core/schematics/migrations/transfer-state/utils.ts:15]:
```suggestion
export const symbolsToUpdate = new Set(['makeStateKey', 'StateKey', 'TransferState']);
```

**Review by AndrewKushnir — COMMENTED** [review]
@JeanMeche we want to try landing this change in v17 and I've left a couple minor comments. Could you please also rebase the PR and resolve merge conflicts?

**Comment by JeanMeche** [comment]
This will need a rebase once #49672 is merged.

File: packages/core/schematics/migrations.json [hunk 8]
```diff
@@ -4,6 +4,11 @@
       "version": "17.0.0",
       "description": "Angular v17 introduces a new control flow syntax that uses the @ and } characters. This migration replaces the existing usages with their corresponding HTML entities.",
       "factory": "./migrations/block-template-entities/bundle"
+    },
+    "migration-v17-transfer-state": {
```
Comment (alan-agius4) [inline packages/core/schematics/migrations.json:8]: NIT 

```suggestion
    "transfer-state": {
```

**Review by AndrewKushnir — APPROVED** [review]
👍

**Comment by AndrewKushnir** [comment]
@crisbeto could you please re-review when you get a chance?

**Review by crisbeto — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
@JeanMeche the PR https://github.com/angular/angular/pull/49672 is now merged. Could you please rebase this PR when you get a chance? Thank you.

**Comment by AndrewKushnir** [comment]
[Presubmit](https://fusion2.corp.google.com/presubmit/571220473/OCL:571220473:BASE:572124157:1696911496517:1a7c08eb/targets).

**Comment by JeanMeche** [comment]
@AndrewKushnir Rebase ✅

**Comment by AndrewKushnir** [comment]
FYI, I've pushed a rebase, since there was a new merge conflict after merging another migration into the `main` branch.

**Comment by AndrewKushnir** [comment]
**Caretaker note**: presubmit is "green", CI failures are unrelated. This PR is ready for merge.

**Pull request closed without merging**